database/src/sqlite3.cpp

```diff
@@ -0,0 +1,121 @@
+#include <database.hpp>
+
+namespace Database
+{
+    Value database_sqlite_open(std::vector<Value>& args, Ark::VM* vm [[maybe_unused]])
+    {
+        if (!types::check(args, ValueType::String))
+            types::generateError(
+                "database:sqlite:open",
+                { { types::Contract { { types::Typedef("database_filename", ValueType::String) } } } },
+                args);
+
+        sqlite3* db;
+        const char* database_name = args[0].string().c_str();
+
+        if (sqlite3_open(database_name, &db) != SQLITE_OK)
+            return Nil;
+
+        return Ark::Value(Ark::UserType(db));
+    }
+
+    Value database_sqlite_close(std::vector<Value>& args, Ark::VM* vm [[maybe_unused]])
+    {
+        if (!types::check(args, ValueType::User) || !args[0].usertype().is<sqlite3>())
+            types::generateError(
+                "database:sqlite:close",
+                { { types::Contract { { types::Typedef("database", ValueType::User) } } } },
+                args);
+
+        sqlite3* db = (sqlite3*)args[0].usertypeRef().data();
+        return Ark::Value(sqlite3_close(db));
+    }
+
+    Value database_sqlite_exec(std::vector<Value>& args, Ark::VM* vm [[maybe_unused]])
+    {
+        if (!types::check(args, ValueType::User, ValueType::String) || !args[0].usertype().is<sqlite3>())
+            types::generateError(
+                "database:sqlite:exec",
+                { { types::Contract { { types::Typedef("database", ValueType::User),
+                                        types::Typedef("sql", ValueType::String) } } } },
+                args);
+
+        sqlite3* db = (sqlite3*)args[0].usertypeRef().data();
+        const char* sql = args[1].string().c_str();
+        char* error_message = nullptr;
+
+        int code = sqlite3_exec(db, sql, nullptr, nullptr, &error_message);
+        if (code != SQLITE_OK)
+        {
+            std::runtime_error error = std::runtime_error(
+                "database:sqlite:exec: exited with code " + std::to_string(code) + "\n" + std::string(error_message));
+            sqlite3_free(error_message);
+            throw error;
+        }
+
+        return Ark::Value(code);
+    }
+
+    Value database_sqlite_exec_with_callback(std::vector<Value>& args, Ark::VM* vm [[maybe_unused]])
+    {
+        if (args.size() != 3 || !args[0].usertype().is<sqlite3>() || args[1].valueType() != ValueType::String || !args[2].isFunction())
```

Comment:
You might want to type check against ValueType::User for args[0], otherwise it would crash if someone gives a number or string, by trying to extract an usertype which isn't there.

Reply:
Nice catch :sunglasses: